storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -96,56 +111,132 @@ func (s *commitLogSource) Read(
+	}
+
+	// Spin up numConc background go-routines to handle M3TSZ encoding. This must
+	// happen before we start reading to prevent infinitely blocking writes to
+	// the encoderChans.
+	wg := &sync.WaitGroup{}
+	for workerNum, encoderChan := range encoderChans {
+		wg.Add(1)
+		go s.startM3TSZEncodingWorker(
+			workerNum,
+			encoderChan,
+			&unmerged,
+			unmergedLock,
+			encoderPool,
+			workerErrs,
+			blopts,
+			wg,
+		)
+	}
+
 	for iter.Next() {
 		series, dp, unit, annotation := iter.Current()
-
-		// check if the series belongs to current namespace being bootstrapped
-		if !namespace.Equal(series.Namespace) {
+		if !s.shouldEncodeSeries(namespace, shardsTimeRanges, blockSize, series, dp) {
 			continue
 		}
 
-		ranges, ok := shardsTimeRanges[series.Shard]
-		if !ok {
-			// Not bootstrapping this shard
-			continue
+		// We use the unmerged slice for lookup because its faster than a map.
+		// Shard ID's map directly to indices, but we don't know the highest shard
+		// number upfront so we resize when necessary. We don't need to acquire a
+		// lock when we check the length because this is the only goroutine that
+		// change the size of the slice.
+		if len(unmerged) < int(series.Shard+1) {
+			// Resize shard
+			unmergedLock.Lock()
+			oldUnmerged := unmerged
+			unmerged = make([]map[ts.Hash]encodersByTime, series.Shard+1)
+			copy(unmerged, oldUnmerged)
+			unmergedLock.Unlock()
 		}
 
-		blockStart := dp.Timestamp.Truncate(blockSize)
-		blockEnd := blockStart.Add(blockSize)
-		blockRange := xtime.Range{
-			Start: blockStart,
-			End:   blockEnd,
-		}
-		if !ranges.Overlaps(blockRange) {
-			// Data in this block does not match the requested ranges
-			continue
+		// Distribute work such that each encoder goroutine is responsible for
+		// approximately numShards / numConc shards. This also means that all
+		// datapoints for a given shard/series will be processed in a serialized
+		// manner.
+		encoderChans[series.Shard%uint32(numConc)] <- struct {
+			series     commitlog.Series
+			dp         ts.Datapoint
+			unit       xtime.Unit
+			annotation ts.Annotation
+			blockStart time.Time
+		}{
+			series:     series,
+			dp:         dp,
+			unit:       unit,
+			annotation: annotation,
+			blockStart: dp.Timestamp.Truncate(blockSize),
 		}
+	}
+	for _, encoderChan := range encoderChans {
+		close(encoderChan)
+	}
 
-		unmergedShard, ok := unmerged[series.Shard]
-		if !ok {
-			unmergedShard = make(map[ts.Hash]encoderMap)
+	// Block until all data has been read and encoded by the worker goroutines
+	wg.Wait()
+	s.printM3TSZEncodingOutcome(workerErrs, iter)
+
+	return s.mergeShards(numShards, bopts, blopts, encoderPool, unmerged), nil
+}
+
+func (s *commitLogSource) startM3TSZEncodingWorker(
+	workerNum int,
+	ec <-chan encoderArg,
+	// Pass as pointer so slice resizing is reflected elsewhere
+	unmergedPtr *[]map[ts.Hash]encodersByTime,
+	unmergedLock *sync.RWMutex,
```

Comment:
If we range over `shardTimeRanges` and allocate the slice up front knowing the largest shard we'll ever encounter, and we also modulo using the current shard for the series then each worker has their own independent set of shards they're decoding for.

Using this approach we can get rid of all locking w.r.t. to unmerged encoders.